### DISCUSSION TITLE
Prep for 3.3.0, 3.4.0, 3.5.0

### DISCUSSION THREAD

**Discussion by mokun** [mars-sim/mars-sim#371] [Ideas]
@bevans2000 

Can we do some project management here, discussing what we want to do in `3.3.x`,  `3.4.x`, and `3.5.x` respectively ?

Say if each of us will work on no more than one or two related minor issues on each of the following official release, what will they be ? 

- `3.3.x`
  -- e.g. #359, 
  -- e.g. #357 

- `3.4.x` 
  -- e.g. #334

- `3.5.x`
  -- e.g. #348 
  -- e.g. #364 

On the other hand, the x is just a patch version that goes after each minor version update, as you know.

So for now, if we patch a few things, we can quickly release `3.2.1`, `3.2.2`, `3.2.3`, and so on.

Shall we ?

**Comment by bevans2000** [comment]
Yes this approach makes sense. Create some Milestones and then we can assign some Issues to them. 

Also think of one big new addition/feature per release as well as improvements/refinements.

**Reply by mokun** [reply]
So, for v3.2.1, how should we update the file header ? 
```

/**
 * Mars Simulation Project
 * SettlementConfig.java
 * @version 3.2.0 2021-07-10
 * @author Barry Evans
 */
```

Update all files from @version 3.2.0 to @version 3.2.1 ?

What about the date ?

**Reply by bevans2000** [reply]
I think we shouldn't be updating all the files just to update the release number. This creates an artificial version for every file even though they is no logical change.

**Reply by mokun** [reply]
IntelliJ complains that we have two asterisks after the slash `/**` and recommends one asterisk after the slash

```
/*
 * Mars Simulation Project
 * SettlementConfig.java
 * @date 2021-08-21
 * @author Barry Evans
 */
```

I don't know there is a java standard for file header. 

But as we have discussed previously, you pointed out it's not that meaningful to have the @version but the date is still good to stay, right ?

So, do you agree that the above header is what it will look like for all newly edited class files ?

**Comment by mokun** [comment]
So for the patch versions, so far we have 
- `3.2.1`
  -- #374
  -- #387
  -- #359 
  -- https://github.com/mars-sim/mars-sim/issues/412

**Reply by mokun** [reply]
@bevans2000 

Is it about time we release `3.2.1` ?

**Reply by bevans2000** [reply]
Yes why not. But #412 is still open, I think you mean #409 which I will close soon. We should close #374.

**Reply by bevans2000** [reply]
Can we close #403 as well and add it?

**Reply by mokun** [reply]
If we no longer want to update the version tag and the date tag on the top of each file, should we go ahead deleting them ? 

```
/**
 * Mars Simulation Project
 * MissionManager.java
 * @version 3.2.0 2021-06-20
 * @author Scott Davis
 */
```

It looks odd to release the next official version, say, 3.2.1 but not changing the version tag or deleting it.

**Reply by mokun** [reply]
May be we can delete the version but still keep the date. 

Each time we edit a file, we simply remind ourselves to manually edit the date, unless there's an automated method out there we can adopt whenever we change the content of a file.

This way, the date will always be current and "up-to-date". No need of changing all the version tags in one commit.

**Reply by bevans2000** [reply]
I see no benefit in deleting the existing header

**Reply by mokun** [reply]
> I see no benefit in deleting the existing header

Not the whole header.

Just the version tag 

but keep the date tag since we no longer update the version tag

**Reply by bevans2000** [reply]
Yes we could remove the version tag but I wouldn't explicitly change all the files to remove the tag. Just remove it whenever we update a file for changes.
Should also look to automatically update the build number when there is a push

**Reply by mokun** [reply]
i just looked up git. we could make change and commit WITHOUT a commit comment. 

This way, if we remove all version tags from all the files and commit WITHOUT a commit comment, the old comment from last change on each file will preserve, since this is your main concern, right ?

I'll test it on a file or two to see what it looks like.

**Reply by mokun** [reply]
So we should just bump up the minor version to make the next release v3.3.0 with.

So far we have 
- `3.3.0`
  -- #374
  -- #387
  -- #359 
  -- #412
  -- #410

**Reply by bevans2000** [reply]
Yes ok. We should do #411 as well since that would give some new functionality..

**Reply by mokun** [reply]
I see your multi-crew selection is already a brand new feature.

#411 would take time. Shall we move it to v3.4.0 ?

But we can certainly fix a bug or two before hitting v3.3.0, right ?

**Reply by bevans2000** [reply]
Ok, sure.

**Reply by mokun** [reply]
Now that you've finished https://github.com/mars-sim/mars-sim/issues/411 and also fixed https://github.com/mars-sim/mars-sim/issues/439, at this point, are we ready to release v3.3.0 yet ?

That way we can move on the next feature.

**Reply by mokun** [reply]
We've addressed 20 issues:
- `3.3.0` 
  -- #81 
  -- #334  
  -- #361
  -- #374
  -- #376
  -- #377
  -- #387
  -- #401
  -- #410
  -- #411
  -- #412
  -- #414
  -- #415
  -- #416
  -- #418
  -- #434
  -- #439
  -- #442
  -- #446
  -- #447
See https://github.com/mars-sim/mars-sim/milestone/6?closed=1

**Reply by bevans2000** [reply]
Sounds a good idea

**Reply by mokun** [reply]
@bevans2000 

Do we expect the beryx console to pop when running the Eclipse launcher "MarsProjectHeadless -new.launch" ?

Also, in a terminal, I run the console edition binary by  
> java -jar 3.3.0_console_java11.jar

Do we expect the beryx console to pop up to give us insight on what's happening with the simulation ?

**Reply by bevans2000** [reply]
If running in headless mode then the beryx console should not appear because that mode means there is no ui screen. 
However, I see no need to keep the headless module anymore because withthe use of arguments the same can be achieved. We would add a new '-noconsole' argument to match the '-nogui' argument.
Having a new '-headless' argument would apply '-nogui -noconsole -remote' to the arguments as a shortcut.

**Reply by bevans2000** [reply]
Also if we apply #452 then the h3wdless module can go.

**Comment by mokun** [comment]
@bevans2000  @Urwumpe 

So this time we'll try something new. 

If we are to patch 3.3.0 by fixing any bugs in 3.3.0, we will call it 3.3.1, 

Should we now create 3.3.1 branch ? Then do a pull from 3.3.1 branch and push it to 3.3.1 branch ?

**Reply by bevans2000** [reply]
We can just use the existing 3.3.0 branch as it is a bugfix of that release. I have logged a bug #460. Do you want that fixing on the branch?

**Reply by mokun** [reply]
> Do you want that fixing on the branch?

Sure. Let's do it on 3.3.0 branch to see how well it fares.

**Reply by mokun** [reply]
Now that we have done a lot of work on refining the storage, we may as well focus on v3.4.0, instead of going backward to release v3.3.1. agree ?

**Reply by bevans2000** [reply]
Yes

**Comment by mokun** [comment]
What else do we still want to take care of before releasing v3.4.0 ?

**Reply by bevans2000** [reply]
I've got some changes for #470 that i am working on involving improving the ```WalkingSteps``` and ```BuildingAirlock```. Removing duplicate code and excessive parameters. I'm also trying to resurrect the walking unit tests. #470 has proved to be more work than I planned but it is flushing out some unclear code. Apart from that I won't be be doing anything else as I'm short of time with work pressure.

**Reply by mokun** [reply]
So. Let's ramp down now. What is a must-fix for you before releasing v3.4.0 ?

**Reply by bevans2000** [reply]
I would like to finish #531 & #509. Plus we have a few Issues that I think can be closed.

**Reply by mokun** [reply]
Should probably include #527 as well.

Will #527 be a fast fix ?

**Reply by bevans2000** [reply]
I'm not sure an the  moment but I will look at this once I've had enough of #531

**Reply by mokun** [reply]
Do you feel stable enough for us to release 3.4.0 ?

**Reply by mokun** [reply]
btw, we haven't identified why the EVA repair of the Lander Hab cannot be done in #509.

**Reply by mokun** [reply]
Let's see if we can fix #509 and #545 before releasing 3.4.0. Shall we ?

**Reply by bevans2000** [reply]
OK, I know nothing about #545. For #509 I recently committed a change to solve the problem with EVAs not starting because it always reports as tooDark.

**Reply by mokun** [reply]
> For #509 I recently committed a change to solve the problem with EVAs not starting because it always reports as tooDark.

So do you mean it is no longer an issue ? Or is it that when it is too dark, it simply prints a log statement ?

**Reply by bevans2000** [reply]
Hmmm, I think I resolved the problem. Certainly don't see any problems at the moment.

**Reply by mokun** [reply]
anything else waiting to be fixed before releasing 3.4.0 ?

**Reply by bevans2000** [reply]
No. I think there are some other issues fixed by #509 but have t check yet. Hoping for some spare time over Easter.

**Reply by mokun** [reply]
So let's close out these open issues for 3.4.0 as much as we can.

**Reply by mokun** [reply]
See https://github.com/mars-sim/mars-sim/milestone/7

![image](https://user-images.githubusercontent.com/1168584/173456072-75e03ef0-7197-4436-b3d8-351950fb571c.png)

**Reply by mokun** [reply]
@bevans2000 

So you think we we've cleared most critical issues and ready for 3.4.0 ? 

We no longer have any outstanding issues using this tracker: 
> https://github.com/mars-sim/mars-sim/milestone/7

btw, I know we did postpone some lingering issues for future versions to deal with.

Are you on course to finish processes.xml as said in https://github.com/mars-sim/mars-sim/issues/729 ?

**Reply by bevans2000** [reply]
Yes but it will take until the weekend due to a shortage of free time. So I'll leave it up to you on whether to close the release or not.

**Reply by bevans2000** [reply]
@mokun  I think we should close 3.4.0 now. Should we just focus on flushing out any lingering problems?
The changes (Shift pattern, Settlement Task board etc) we are discussing feel too big to keep squeezing into 3.4.0.

**Reply by mokun** [reply]
> @mokun I think we should close 3.4.0 now.

Done. We can work on either 3.4.1 or 3.5.0 now.

**Reply by bevans2000** [reply]
We should work on 3.5.0. 3.4.1 is for a patch release if we need one.

**Reply by mokun** [reply]
@bevans2000 

I made a mistake by choosing "Push to Origin" in Eclipse IDE. 

As a result, this causes all the recent changes to be merged with 3.4.0 branch. 

What I want was simply commit some changes to the Master branch. But I don't know why this happens.

How do you get rid of those changes that are NOT for 3.4.0 branch ?

Is there a way to reverse all those commits that were done for pre-3.5.0 ?

**Reply by bevans2000** [reply]
Yes it can be fixed. I can look at weekend, it's just a case of rolling back the branch to the 3.4.0 tag. Are all your commits on the master branch?

**Reply by mokun** [reply]
>  it's just a case of rolling back the branch to the 3.4.0 tag.

Ok. See https://github.com/mars-sim/mars-sim/commits/3.4.0

The first 3.5.0 commit should be [this](https://github.com/mars-sim/mars-sim/commit/5b32d4febb315481575f6711e291fda372a348f0)

Everything else before that should be included in 3.4.0.

> Are all your commits on the master branch?

Yes. I just did [this](https://github.com/mars-sim/mars-sim/commit/91ab44d20da2e9dc46f096089dd4d7728c385749)

**Reply by bevans2000** [reply]
Ok, it can be fixed, just need to look at it on my laptop.

**Reply by bevans2000** [reply]
Easiest approach is I will delete the 3,4.0 remote branch and just recreate it from  the tag. I have removed the old branches as well because we will not create a patch release for those now. Key is to always tag the code then you can always create the branches later.

**Reply by bevans2000** [reply]
So I've deleted the 3.4.0 branch but not recreated it yet. Usually you only create a bug fix branch if you need to create a bug fix. The releases are tagged so we can always create a branch from the tag quite easily.
Also don't create the branch with the same name as the tag because it can confuse git.

**Reply by mokun** [reply]
> Usually you only create a bug fix branch if you need to create a bug fix.

ok.

>  The releases are tagged so we can always create a branch from the tag quite easily.

Do you mean that one can recreate a branch from the tags at https://github.com/mars-sim/mars-sim/tags ?

**Reply by mokun** [reply]
> Also don't create the branch with the same name as the tag because it can confuse git.

Sure. In future, I'll add a "v" before the release number on either the tag or the branch so that they are different.

**Reply by bevans2000** [reply]
> > Usually you only create a bug fix branch if you need to create a bug fix.
> 
> ok.
> 
> > The releases are tagged so we can always create a branch from the tag quite easily.
> 
> Do you mean that one can recreate a branch from the tags at https://github.com/mars-sim/mars-sim/tags ?

Yes you can checkout the tag directly into a new branch. https://devconnected.com/how-to-checkout-git-tags/#:~:text=In%20order%20to%20checkout%20a,branch%20to%20be%20checked%20out.&text=Note%20that%20you%20will%20have,list%20from%20your%20remote%20repository.

**Comment by mokun** [comment]
So for 3.5.0, what do we want to accomplish ? 

May be we should bump up more quickly and limit each minor version to have no more than three major improvement. Shall we ?

**Reply by bevans2000** [reply]
Yes I think we sat on 3.4.0 too long. What about time boxing it? To a major release every 6 months? The last month focus on stability. Big changes done in the first 3 months

**Reply by mokun** [reply]
> To a major release every 6 months?

when you say major release, do you mean bumping up the first digital, say, from 3.4.0 to 4.0.0 ?

**Reply by bevans2000** [reply]
No sorry, I meant minor.

**Reply by mokun** [reply]
@bevans2000 So when do we release v3.4.1 ?

**Reply by bevans2000** [reply]
It's done. I did it last week when I fixed the start up problem

**Reply by mokun** [reply]
Sorry I meant v3.5.0. what will be the major fixes ?

**Reply by bevans2000** [reply]
Do we want to move to Java 17? As the current release doesn't work on the latest LTS I think that's a big problem. So do you want to fix that as a patch 3 4.2 or do it in 3.5.0 or do it even later?
Nevertheless we should be using the 3.5.0 Milestone to drive when 3.5.0 is ready.

**Reply by Urwumpe** [reply]
I would say, it should be patched ASAP, including the risk that the patch could cause technical debt, which would then be scheduled to be fixed until the next major version. 

Not running on the current LTS after over a year has passed is not very honorable, but I know large companies* still fighting with the migration to Java 8, so we should at least try our best there to fix it. The standard support for Java 11 will end this year, it should be done by that time so we also have the capacity to look into compatibility with the upcoming Java 21 LTS.

**Reply by bevans2000** [reply]
I agree that we need to do this soon. I am thinking we target 3.5.0 as the transition off Weblaf is big.

**Reply by mokun** [reply]
> So do you want to fix that as a patch 3.4.2

What are to be included in 3.4.2 ?

**Reply by bevans2000** [reply]
The fix for Java 17 as including #789 but it's too large an activity for a bugfix release. We have to move dropping weblaf to 3.5 is my opinion.

**Reply by Urwumpe** [reply]
Then we should try providing a start script with the needed module exceptions for WebLAF. I am not sure if everything works and so many exceptions makes me feel terribly bad, but maybe its a fix until we replaced WebLAF or WebLAF learned Java 17.

**Reply by bevans2000** [reply]
Should be close 3.5.0 now? There have been many changes in terms of code/design improvements as well as new features. I'm close to closing down the Swing improvement work.
#306 is on the 3.5.0 milestone but this needs a wide ranging change and feels too big to add to the current release.
Do we want to jump to Java 17 as part of 3.5.0 ? Dropping WebLAF should open this up now.

**Reply by mokun** [reply]
Yay we may as well switch to Java 17 when releasing 3.5.0.

@Urwumpe anything else you see it as necessary to fix in 3.5.0 ?

**Reply by Urwumpe** [reply]
@mokun Only a high casualty rate during the last test run of about one death per sol, but I believe most of them are known issues already. No showstoppers for the next release.

**Reply by mokun** [reply]
> most of them are known issues already.

such as ?

**Reply by Urwumpe** [reply]
For example a lot of those resulting in Settlers entering an infinite loop: 

```
01-Adir-04:683,982 (Info) ExitAirlock : [New Plymouth - EVA Airlock 1] George Clark - Not fit enough to depressurize chamber.
01-Adir-04:707,208 (Info) ExitAirlock [x133] : [New Plymouth - EVA Airlock 1] George Clark - Airlock is not in egress mode. Current task: Performing EVA Egress.
01-Adir-04:730,378 (Info) ExitAirlock [x134] : [New Plymouth - EVA Airlock 1] George Clark - Airlock is not in egress mode. Current task: Performing EVA Egress.
01-Adir-04:753,573 (Info) ExitAirlock [x134] : [New Plymouth - EVA Airlock 1] George Clark - Not fit enough to request egress. Current task: Performing EVA Egress.
```

Another smaller issue observed:

```
01-Adir-04:750,306 (Warning) Task [x3] : [Starcity] DeliveryBot 002 - Early end of Unloading vehicle: LUV 012 no garage found
01-Adir-04:751,522 (Warning) Task [x4] : [Starcity] DeliveryBot 001 - Early end of Unloading vehicle: LUV 012 no garage found
```

**Reply by mokun** [reply]
> George Clark - fit enough to ....

I'd say this log is normal. It simply says a person may have a lot of fatigue and is not suited for EVA. It prints at what task phase the person is barred from continuing the egress.

> George Clark - Airlock is not in egress mode. Current task: Performing EVA Egress.

The EVA airlock is either OFF or in ingress mode. Therefore, George can't go out and will have to wait.

**Reply by mokun** [reply]
@bevans2000 

> DeliveryBot 002 - Early end of Unloading vehicle: LUV 012 no garage found

Can you comment on this ?

**Reply by Urwumpe** [reply]
> > George Clark - fit enough to ....
> 
> I'd say this log is normal. It simply says a person may have a lot of fatigue and is not suited for EVA. It prints at what task phase the person is barred from continuing the egress.

But that means he is stuck in a task that he shouldn't be doing in first place. 

I would expect him to get some rest or medical care, then and be removed from regular duty for a while.

> 
> > George Clark - Airlock is not in egress mode. Current task: Performing EVA Egress.
> 
> The EVA airlock is either OFF or in ingress mode. Therefore, George can't go out and will have to wait.

The problem is, that he is still trying to enter the airlock for egresss after being prohibited to do it 50 millisols (over one hour) earlier. Better would be fixing the cause, not trying again until the settler breaks. I think the check that somebody is unable to do a EVA related task happens too late then. 

Defining such rules for any EVA would be easier if we could include some rules or even rulebooks applying to a task (like AOP without being AOP) and have something like exception handling available. So that general "EVA mission rules" can apply to a task (Like forbidding certain health issues, bad EVA suit state, etc.. all the time during an EVA), but how a violated rule is handled depends on the situation by defining an "exception handler" there, maybe just by switching to an emergency action.

And can I have a button or console command to extract a log of the previous actions done by a settler into a text file? That could make such analyses easier.

**Reply by bevans2000** [reply]
> And can I have a button or console command to extract a log of the previous actions done by a settler into a text file? That could make such analyses easier.

Yes, depends what level of details we want. If it's just the level of the current Activity Tab then that's easy but I wonder if in this case we actually need a more detailed view. The detailed view would have to be for a shorter period than the current 5 sols.

**Reply by bevans2000** [reply]
> The problem is, that he is still trying to enter the airlock for egresss after being prohibited to do it 50 millisols (over one hour) earlier. Better would be fixing the cause, not trying again until the settler breaks. I think the check that somebody is unable to do a EVA related task happens too late then.
> 
> Defining such rules for any EVA would be easier if we could include some rules or even rulebooks applying to a task (like AOP without being AOP) and have something like exception handling available. So that general "EVA mission rules" can apply to a task (Like forbidding certain health issues, bad EVA suit state, etc.. all the time during an EVA), but how a violated rule is handled depends on the situation by defining an "exception handler" there, maybe just by switching to an emergency action.

This is a common problem of the airlocks getting overloaded. I feel the cycle time is too long although I've no scientific evidence on this :-)
The Persons are healthy when they start the task but because they wait for so long, they become unhealthy. There does seem to be an underlying problem with the queuing for al airlock.

**Reply by bevans2000** [reply]
> @bevans2000
> 
> > DeliveryBot 002 - Early end of Unloading vehicle: LUV 012 no garage found
> 
> Can you comment on this ?

A Robot has tried to start an Unlock tsk but there is no Garage space. What is interesting is the Vehicle is a LUV which means a Construction Mission must have been done. I've never seen this scenario, how did it occur?

**Reply by mokun** [reply]
> The problem is, that he is still trying to enter the airlock for egresss after being prohibited to do it 50 millisols (over one hour) earlier. Better would be fixing the cause, not trying again until the settler breaks. I think the check that somebody is unable to do a EVA related task happens too late then.

Yes agreed. I have tried to make the Walk related classes return a boolean value in such a way that it will make it impossible to even start the EVA egress. But it will need a lot of changes.

> Defining such rules for any EVA would be easier if we could include some rules or even rulebooks applying to a task (like AOP without being AOP) and have something like exception handling available. 

What's AOP ?

So that general "EVA mission rules" can apply to a task (Like forbidding certain health issues, bad EVA suit state, etc.. all the time during an EVA), but how a violated rule is handled depends on the situation

I've tried to create an airlock reservation queue but it wasn't successful. 

May be your EVA mission ruleset could be a good approach and we can discuss more on how to implement it

**Comment by bevans2000** [comment]
I have created an issue for the Java17 upgrade #820. Do we want to do this before closing 3.5.0 or push it to 3.6.0?
I have a couple of changes left on #810 but once completed 3.5.0 is done in my mind.

**Reply by mokun** [reply]
Yes. I suppose we can switch to Java 17 now. 

Perhaps one last question I have is whether the current version of kotlin is compatible with Java 17.

**Reply by bevans2000** [reply]
> Yes. I suppose we can switch to Java 17 now.
> 
> Perhaps one last question I have is whether the current version of kotlin is compatible with Java 17.

Yes https://discuss.kotlinlang.org/t/using-kotlin-1-6-and-java-17-in-compose-multiplatform/23722

**Comment by bevans2000** [comment]
Good question. Do we actually need kotlin?

**Reply by mokun** [reply]
> Good question. Do we actually need kotlin?

To me, it serves two purposes. 

1. I want to use it to program the use of neural network. 
2. FXGL uses kotlin for writing its source codes. If kotlin is installed, I will be able to look into how the API works internally. 

So I hope to keep that. 

Also, more and more JavaFX libraries out there are using kotlin.

**Reply by bevans2000** [reply]
JavaFX is separate now so does that mean users need an extra download to use the app ? https://blogs.oracle.com/java/post/the-future-of-javafx-and-other-java-client-roadmap-updates

**Reply by mokun** [reply]
> JavaFX is separate now so does that mean users need an extra download to use the app ? 

No. The FXGL maven artifact automatically takes care of selecting and downloading a version of JavaFX that it can use.

Therefore, as long as the mars-sim-fxgl submodule is being used, users won't need to download JavaFX separately and won't need to install it manually :D

**Comment by bevans2000** [comment]
@mokun When are we going to close down 3.5.0? Java 17 is done. I'm not making any more changes under 3.5.0 now. I think we just need to check the current master code base with some runs.

**Reply by mokun** [reply]
I just uploaded the 3.4.1 to sourceforge.

Do you want to annotate the changes/fixes in Changelog ?

We can release it tomorrow.

**Reply by bevans2000** [reply]
Wow, do we still bother with SourceForge?

**Reply by bevans2000** [reply]
Let's make the 3.5 release. If we find any other problems I'll fix them in a bugfix branch.

**Reply by mokun** [reply]
ok. I'm going to make the last commit to update the dates right now and add a new branch.

**Reply by mokun** [reply]
done.

Can you create a 3.5.0 branch now ?

**Reply by bevans2000** [reply]
As long as the release is tagged we can create a branch when we need it

**Reply by mokun** [reply]
> As long as the release is tagged we can create a branch when we need it

Yes. It needs to `choose a tag` and this tag needs to be a branch other than Master.

I don't have the git command line anymore and I forgot how to create a new brand in eclipse.

Can you create 3.5.0 now ?

Then you or I will create a new entry as the official release.

**Reply by bevans2000** [reply]
Sure I can create it.

**Reply by mokun** [reply]
Btw I've posted the 3.5.0 release.

What's next now ?

**Reply by bevans2000** [reply]
Change the pom files to be version 3.6.0?

**Reply by mokun** [reply]
Sure. Should we change them to post-3.5.0 or pre-3.6.0 instead ?

**Reply by bevans2000** [reply]
It's done and I have started creating a page to the Wiki detailing the steps for a release. Do you want to update it with anything I have missed.
https://github.com/mars-sim/mars-sim/wiki/Release-Procedure

**Reply by mokun** [reply]
> Wiki detailing the steps for a release.

That's thoughtful ! 

It's certainly good to have this wiki. Not only is it great for the continuity of this open source project, but also for being thorough in formalizing and documenting exactly what we do when publishing an official release.

**Reply by mokun** [reply]
I've just edited this wiki.

btw, we should update the [Roadmap](https://github.com/mars-sim/mars-sim/wiki/Roadmap). It was woefully outdated.

**Reply by bevans2000** [reply]
Yes I thought it makes to document the steps. Ideally we should automate the changes with a GitFlow action but that might be wishful thinking.
